docs/decisions/0035-embedding-abstraction.md

```diff
@@ -0,0 +1,257 @@
+
+### Embeddings models
+
+OpenAI provides models that allow for generating embeddings only from textual data.\
+The models do not return metadata, and also do not allow for parameterizing the query.
+
+Google VertexAI provides models that allow for generating embeddings from textual data, as well as from images and other types of data.
+The models return metadata, and also allow for parameterizing the query.
+
+<!-- This is an optional element. Feel free to remove. -->
+
+## Decision Drivers
+
+1. Abstraction should be able to generate embeddings from different types of data, at least from text and images.
+2. Abstraction should return metadata with generated embeddings.
+3. Abstraction should allow parameterize the embeddings query.
+4. We should maintain the precision of the obtained embeddings.
+
+## Considered Options
+
+### Option 1 [Current] - Generic embeddings generation interface and specialized interfaces for different types of data
+
+The current abstraction of generating embeddings only allows for generation from text and image data.\
+We return raw data in the form of `IList<ReadOnlyMemory<TEmbedding>>` instead of a specialized data type like `IList<EmbeddingContent>`.
+
+In this option, we cannot parameterize the query or return metadata.
+
+```csharp
+public interface IEmbeddingGenerationService<TValue, TEmbedding> : IAIService where TEmbedding : unmanaged
+{
+    Task<IList<ReadOnlyMemory<TEmbedding>>> GenerateEmbeddingsAsync(
+        IList<TValue> data,
+        Kernel? kernel = null,
+        CancellationToken cancellationToken = default);
+}
+
+public interface ITextEmbeddingGenerationService : IEmbeddingGenerationService<string, float> { }
+
+public interface IImageEmbeddingGenerationService : IEmbeddingGenerationService<ImageContent, float> { }
+```
+
+Pros:
+- Allows for generating embeddings from different types of data.
+- Generic interface.
+- Interface segregation.
+
+Cons:
+- Cannot parameterize the query.
+- Cannot return metadata.
+- We lost precision.
+
+### Option 2 [Proposed] - Specialized embeddings generation interfaces for different types of data with metadata and query parameterization without a generic interface
+
+This option allows you to generate embeds from different data types, such as text and images.\
+And allows you to parameterize the query and return metadata.\
+We can also develop interfaces with other returned types like `EmbeddingContent<int>`, `EmbeddingContent<double>`, etc.
+
+```csharp
+public class EmbeddingContent<TEmbedding> : KernelContent where TEmbedding : unmanaged
+{
+    public EmbeddingContent(
+        IReadOnlyList<ReadOnlyMemory<TEmbedding>> data,
+        string? modelId = null,
+        object? innerContent = null,
+        IReadOnlyDictionary<string, object?>? metadata = null)
+        : base(innerContent, modelId, metadata)
+    {
+        this.Data = data;
+    }
+
+    public IReadOnlyList<ReadOnlyMemory<TEmbedding>> Data { get; set; }
```

Comment:
In `EmbeddingContent` class `Data` property is `IReadOnlyList` but then in all provided options, interface method `GenerateEmbeddingsAsync` returns `IReadOnlyList<EmbeddingContent<TEmbedding>>`, so it's collection of collection of embedding. Not sure if it's needed. 

In `EmbeddingContent` class we can keep `Data` as `ReadOnlyMemory<TEmbedding>` and in interface return `IReadOnlyList<EmbeddingContent<TEmbedding>>`, which would be multiple responses with its own vector and metadata. I think it should be enough.

Reply:
yeah, I did not notice my error 🙃